### PR TITLE
Honor nullable params in Kotlin

### LINKED ISSUE #397: nullable params with Kotlin
## Overview

When the function annotated with `@NeedsPermission` has a nullable parameter, the generated function in the dispatcher have a non nullable one instead.

### Expected

The `**WithPermissionCheck` generated method should use a nullable parameter too

### Actual

The `**WithPermissionCheck` use a non-nullable parameter

## Environment

- v3.0.1

## Reproducible steps

```kotlin
//Function in Activity
@NeedsPermission(Manifest.permission.ACCESS_COARSE_LOCATION)
internal fun functionWithNullableParameter(string: String?) {
    //...
}

//Generated function in the PermissionsDispatcher
fun HomeActivity.functionWithNullableParameterWithPermissionCheck(string: String) {
    //...
}

```

### PR DESCRIPTION
**Pull request by mannodermaus**
Generated Kotlin code needs to honor nullable parameters in the signatures of annotated `@NeedsPermission` methods:

```kotlin
@RuntimePermissions
class SomeActivity : AppCompatActivity() {
  @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
  fun updateLocation(currentLocation: LatLng?) {
    // ...
  }
}

// Wrong
fun SomeActivity.updateLocationWithPermissionCheck(currentLocation: LatLng) {
}

// Correct
fun SomeActivity.updateLocationWithPermissionCheck(currentLocation: LatLng?) {
}
```

As part of the Parameter-to-TypeName chain, we now inspect the annotations of a `VariableElement` & attach the nullability information to its `TypeName` via a new extension method, `mapToNullableTypeIf(Boolean)`. Since we use this chain in multiple places, I have extracted this mapping into its own function, `VariableElement#asPreparedType()`.

This fixes #397.

### CODE REVIEW

**Comment by hotchemi** [comment]
Thx! Seems good but let me have a double check.

File: processor/src/main/kotlin/permissions/dispatcher/processor/impl/ProcessorUnits.kt [hunk 1]
```diff
@@ -6,13 +6,14 @@ import permissions.dispatcher.processor.impl.java.JavaSupportFragmentProcessorUn
 import permissions.dispatcher.processor.impl.kotlin.KotlinActivityProcessorUnit
 import permissions.dispatcher.processor.impl.kotlin.KotlinNativeFragmentProcessorUnit
 import permissions.dispatcher.processor.impl.kotlin.KotlinSupportFragmentProcessorUnit
+import javax.annotation.processing.Messager
 
-fun javaProcessorUnits() = listOf(
-        JavaActivityProcessorUnit(),
-        JavaSupportFragmentProcessorUnit(),
-        JavaNativeFragmentProcessorUnit())
+fun javaProcessorUnits(messager: Messager) = listOf(
```
Comment (shiraji) [inline processor/src/main/kotlin/permissions/dispatcher/processor/impl/ProcessorUnits.kt]: What is the purpose of having `Messager`? Is it for debugging purpose?

Comment (mannodermaus) [reply processor/src/main/kotlin/permissions/dispatcher/processor/impl/ProcessorUnits.kt]: Yeah, I propagated it to the `ProcessorUnit` so I could print some debug statements while developing!

File: processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt [hunk 2]
```diff
@@ -99,5 +115,19 @@ fun FileSpec.Builder.addTypes(types: List<TypeSpec>): FileSpec.Builder {
  * To avoid KotlinPoet bug that returns java.lang.String when type name is kotlin.String.
  * This method should be removed after addressing on KotlinPoet side.
  */
-fun TypeName.checkStringType() =
-        if (this.toString() == "java.lang.String") ClassName("kotlin", "String") else this
+fun TypeName.mapPrimitivesToKotlinTypes() =
+        when (this.toString()) {
+            "java.lang.String" -> ClassName("kotlin", "String")
```
Comment (hotchemi) [inline processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: I think primitive types except `String` are well-converted in [KotlinPoet](https://github.com/square/kotlinpoet/blob/2bca828f131c7f2274246e91639164abfb7a9fb3/src/main/java/com/squareup/kotlinpoet/TypeName.kt#L124). Are there any necessity to take care?

Comment (mannodermaus) [reply processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: When it came to nullable primitives, I remember that the generated code favored `@Nullable java.lang.Integer` over `kotlin.Int?`, which was weird. Maybe it's an issue with KotlinPoet? I'll try to reproduce and open a ticket

Comment (hotchemi) [reply processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: Oh really? I thought that was only a problem with `java.lang.String`? 👀 
I haven't tested in all primitive types so not sure but if the convert is weird it's ok to keep for sure 🙇 
ref: https://github.com/square/kotlinpoet/issues/236

Comment (hotchemi) [reply processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: Sorry for the late, any update for this? 👁

Comment (mannodermaus) [reply processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: Not yet, sorry! I'll try to find a time slot to verify this behavior

Comment (hotchemi) [reply processor/src/main/kotlin/permissions/dispatcher/processor/util/Extensions.kt]: No problem 🙇

**Comment by mannodermaus** [comment]
I wasn't able to repro the "Kotlin primitive mapping" issue anymore, so I have reverted the changes to `checkStringType()` for now. Sorry it took so long! 🙇

**Comment by hotchemi** [comment]
@mannodermaus no problem! thx for your dedicated work! 🙇

**Pull request merged**